test/test_forest_aws.py

```diff
@@ -0,0 +1,124 @@
+                                                           local_path)
+
+
+class TestS3Mount(unittest.TestCase):
+    def setUp(self):
+        self.file_name = "file.nc"
+        self.directory = "directory"
+        self.mount = forest.aws.S3Mount(self.directory)
+
+    def test_mount_directory(self):
+        self.assertEqual(self.mount.directory, self.directory)
+
+    @unittest.mock.patch("forest.aws.os")
+    def test_file_exists_queries_file_system(self, os):
+        self.mount.file_exists(self.file_name)
+        os.path.isfile.assert_called_once_with(self.file_name)
+
+    def test_path_to_load(self):
+        result = self.mount.path_to_load(self.file_name)
+        expect = os.path.join(self.directory, self.file_name)
+        self.assertEqual(result, expect)
+
+
+class TestS3Bucket(unittest.TestCase):
+    """AWS S3 architecture
+
+    Forest application should run from either file system
+    or AWS download seemlessly. Both systems should present
+    a simple API to the internals of Forest
+    """
+    def setUp(self):
+        self.file_name = "file.nc"
+        self.server_address = "https://s3.eu-west-2.amazonaws.com"
+        self.bucket_name = "stephen-sea-public-london"
+        self.s3_base = "{server}/{bucket}/model_data/".format(server=self.server_address,
+                                                              bucket=self.bucket_name)
+        self.download_dir = "download_directory"
+        self.bucket = forest.aws.S3Bucket(self.server_address,
+                                          self.bucket_name,
+                                          self.download_dir)
+
+    def test_server_address(self):
+        self.assertEqual(self.bucket.server_address, self.server_address)
+
+    def test_bucket_name(self):
+        self.assertEqual(self.bucket.bucket_name, self.bucket_name)
+
+    def test_path_to_load(self):
+        result = self.bucket.path_to_load(self.file_name)
+        expect = os.path.join(self.download_dir, self.file_name)
+        self.assertEqual(result, expect)
+
+@unittest.mock.patch("forest.aws.os")
+@unittest.mock.patch("forest.aws.urllib")
+@unittest.mock.patch("forest.aws.print", autospec=True)
+class TestS3BucketIO(unittest.TestCase):
+    def setUp(self):
+        self.file_name = "file.nc"
+        self.server_address = "server_address"
+        self.bucket_name = "bucket_name"
+        self.download_dir = "download_directory"
+        self.bucket = forest.aws.S3Bucket(self.server_address,
+                                          self.bucket_name,
+                                          self.download_dir)
+
+    def test_file_exists_calls_s3_file_exists(self,
+                                              mock_print,
+                                              urllib,
+                                              os):
+        self.bucket.s3_file_exists = unittest.mock.Mock()
+        self.bucket.file_exists(self.file_name)
```

Comment:
I'm not keen on testing the internal workings of a class, only it's public interface. So rather than does it call `self. s3_file_exists` I'd test does it return False if urllib returns an 404.